source/docs/software/networktables/listening-for-change.rst

```diff
@@ -1,57 +1,308 @@
 
-            //add an entry listener for changed values of "Y", the lambda ("->" operator)
-            //defines the code that should run when "Y" changes
-            yEntry.addListener(event -> {
-               System.out.println("Y changed value: " + event.getEntry().getValue());
-            }, EntryListenerFlags.kNew | EntryListenerFlags.kUpdate);
+The below example listens to various kinds of events using ``NetworkTableInstance``. The listener callback provided to any of the addListener functions will be called asynchronously from a background thread when a matching event occurs.
 
-            try {
-               Thread.sleep(10000);
-            } catch (InterruptedException ex) {
-               System.out.println("Interrupted");
-               Thread.currentThread().interrupt();
-               return;
+.. warning:: Because the listener callback is called from a separate background thread, it's important to use thread-safe synchronization approaches such as mutexes or atomics to pass data to/from the main code and the listener callback function.
+
+The ``addListener`` functions in NetworkTableInstance return a listener handle. This can be used to remove the listener later.
+
+.. tabs::
+
+    .. group-tab:: Java
+
+        .. code-block:: java
+
+            public class Example {
+              final DoubleSubscriber ySub;
+              // use an AtomicReference to make updating the value thread-safe
+              final AtomicReference<> yValue = new AtomicReference<Double>();
+              // retain listener handles for later removal
+              int connListenerHandle;
+              int valueListenerHandle;
+              int topicListenerHandle;
+
+              public Example() {
+                // get the default instance of NetworkTables
+                NetworkTableInstance inst = NetworkTableInstance.getDefault();
+
+                // add a connection listener; the first parameter will cause the
+                // callback to be called immediately for any current connections
+                connListenerHandle = inst.addConnectionListener(true, event -> {
+                  if (event.is(NetworkTableEvent.Kind.kConnected)) {
+                    System.out.println("Connected to " + event.connInfo.remote_id);
+                  } else if (event.is(NetworkTableEvent.Kind.kDisconnected)) {
+                    System.out.println("Disconnected from " + event.connInfo.remote_id);
+                  }
+                });
+
+                // get the subtable called "datatable"
+                NetworkTable datatable = inst.getTable("datatable");
+
+                // subscribe to the topic in "datatable" called "Y"
+                ySub = datatable.getDoubleTopic("Y").subscribe(0.0);
+
+                // add a listener to only value changes on the Y subscriber
+                valueListenerHandle = inst.addListener(
+                    ySub,
+                    EnumSet.of(NetworkTableEvent.kValueAll),
```

Comment:
While updating our 2022 code to 2023 wpilib beta 3, I referenced this example. I had to specify NetworkTableEvent.Kind.kValueAll since kValueAll is in the Kind enum.